File: lab-1/src/lab1_pkg/src/talker.cpp
#include "rclcpp/rclcpp.hpp"
#include "ackermann_msgs/msg/ackermann_drive_stamped.hpp"

class TalkerNode : public rclcpp::Node {
public:
    TalkerNode() : Node("talker") {
        this->declare_parameter<double>("v", 0.0);
        this->declare_parameter<double>("d", 0.0);

        publisher_ = this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>("drive", 10);

        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(0),
            std::bind(&TalkerNode::timer_callback, this)
        );
    }

private:
    void timer_callback() {
        double v = this->get_parameter("v").as_double();
        double d = this->get_parameter("d").as_double();

        auto message = ackermann_msgs::msg::AckermannDriveStamped();
        message.drive.speed = v;
        message.drive.steering_angle = d;

        publisher_->publish(message);
		// RCLCPP_INFO(this->get_logger(), "Publishing to /drive: Speed: %f, Steering Angle: %f", v, d);
    }

    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr publisher_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TalkerNode>());
    rclcpp::shutdown();
    return 0;
}

File: lab-1/src/lab1_pkg/scripts/talker.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from ackermann_msgs.msg import AckermannDriveStamped

class TalkerNode(Node):
    def __init__(self):
        super().__init__('talker')
        self.declare_parameter('v', 0.0)
        self.declare_parameter('d', 0.0)

        self.publisher_ = self.create_publisher(AckermannDriveStamped, 'drive', 10)
        timer_period = 0
        self.timer = self.create_timer(timer_period, self.timer_callback)

    def timer_callback(self):
        v = self.get_parameter('v').get_parameter_value().double_value
        d = self.get_parameter('d').get_parameter_value().double_value

        msg = AckermannDriveStamped()
        msg.drive.speed = v
        msg.drive.steering_angle = d
        self.publisher_.publish(msg)
        # self.get_logger().info(f'Publishing to /drive: Speed: {v}, Steering Angle: {d}')

def main(args=None):
    rclpy.init(args=args)
    talker = TalkerNode()
    rclpy.spin(talker)
    rclpy.shutdown()

if __name__ == '__main__':
    main()